lib/espnow.toit

```diff
@@ -0,0 +1,75 @@
+// Copyright (C) 2022 Toitware ApS. All rights reserved.
+// Use of this source code is governed by an MIT-style license that can be
+// found in the lib/LICENSE file.
+
+import bytes
+
+STATION ::= 0 // Wi-Fi station mode
+SOFTAP  ::= 1 // Not support yet
+
+BROADCAST_MAC ::= #[0xff, 0xff, 0xff, 0xff, 0xff, 0xff]
+
+check_mac address/ByteArray:
+  if address.size != 6:
+      throw "ESP-Now MAC address length must be 6 bytes"
+
+check_key key/ByteArray:
+  if key.size != 16:
+      throw "ESP-Now key length must be 16 bytes"
+
+check_channel channel/int:
+  if channel > 14 or channel < 0:
+      throw "ESP-Now channel range must be 0~14"
+
+addr_to_string raw_/ByteArray:
+  buffer := bytes.Buffer
+  6.repeat:
+    if it > 0: buffer.write_byte ':'
+    byte := raw_[it]
+    buffer.write_byte (to_lower_case_hex byte >> 4)
+    buffer.write_byte (to_lower_case_hex byte & 0xf)
+  return buffer.to_string
+
+class Datagram:
+  addr/ByteArray := ?
+  data/ByteArray := ?
+
+  constructor .addr .data:
+
+class Service:
+  _group ::= ?
+
+  constructor --mode/int=STATION --pmk/ByteArray:
+    check_key pmk
+    _group = espnow_init_ mode pmk
+
+  send --addr/ByteArray --data/ByteArray --wait/bool=true -> int:
+    check_mac addr
+    return espnow_send_ addr data wait
+
+  receive -> any:
+    array := espnow_receive_ (Array_ 2)
+    if not array:
+      return null
+    return Datagram array[0] array[1]
+
+  add_peer --addr/ByteArray --channel/int --encrypted/bool --key/ByteArray -> bool:
```

Comment:
The peer is identified by its address, as such it's the main argument and probably doesn't need any named argument.
```suggestion
  add_peer address/ByteArray --channel/int --encrypted/bool --key/ByteArray -> bool:
```

I would combine this and the next function:
```suggestion
  add_peer address/ByteArray --channel/int --key/ByteArray?=null --encrypted/bool=(key != null) -> bool:
    if not key: key = #[]
```

From what I know, the espnow protocol has a standard key it can use. If that's a good default, then I would make `--encrypted` by default equal to true. This is really a question of common use and how most people do/should work with espnow.

Reply:
This key is used to encrypt LMK(Local Master Keys), description is [here](https://docs.espressif.com/projects/esp-idf/en/v4.4.2/esp32/api-reference/network/esp_now.html#security): LMK of the paired device is used to encrypt the vendor-specific action frame with the CCMP method. The maximum number of different LMKs is six. If the LMK of the paired device is not set, the vendor-specific action frame will not be encrypted.
So I think for safety we should make it true and use a default key in this example.

Reply:
Why here add `if not key: key = #[]`, can it pass `key=null` to primitive ?

Reply:
Yes. You can pass null to primitives as well.
You would need to pay attention with the typing of the C++ code, as you wouldn't be able to take a `Blob` (or whatever type you expect) anymore.
I'm fine either way. I believe we canonicalize the #[], so it wouldn't actually cost any allocation.

Reply:
OK, it looks like better to using #[].